packages/react/src/utils/useScrollLock.ts

```diff
@@ -8,172 +9,80 @@ let originalBodyStyles = {};
-      top: htmlStyle.top,
-      left: htmlStyle.left,
-      right: htmlStyle.right,
-      overflowX: htmlStyle.overflowX,
-      overflowY: htmlStyle.overflowY,
+      overflowY: html.style.overflowY,
+      overflowX: html.style.overflowX,
     };
+
     originalBodyStyles = {
-      overflowX: bodyStyle.overflowX,
-      overflowY: bodyStyle.overflowY,
+      position: body.style.position,
+      height: body.style.height,
+      boxSizing: body.style.boxSizing,
+      overflowY: body.style.overflowY,
+      overflowX: body.style.overflowX,
     };
 
+    // Handle `scrollbar-gutter` in Chrome when there is no scrollable content.
+    const hasScrollbarGutterStable = htmlStyles.scrollbarGutter?.includes('stable');
+
     const isScrollableY = html.scrollHeight > html.clientHeight;
     const isScrollableX = html.scrollWidth > html.clientWidth;
+    const hasConstantOverflowY =
+      htmlStyles.overflowY === 'scroll' || bodyStyles.overflowY === 'scroll';
+    const hasConstantOverflowX =
+      htmlStyles.overflowX === 'scroll' || bodyStyles.overflowX === 'scroll';
 
-    // Handle `scrollbar-gutter` in Chrome when there is no scrollable content.
-    const hasScrollbarGutterStable = htmlComputedStyles.scrollbarGutter?.includes('stable');
-
-    // Safari needs visual viewport offsets added to account for pinch-zoom
-    const webkit = isWebKit();
-    const { x, y } = getVisualOffsets(doc);
-    const visualX = webkit ? x : 0;
-    const visualY = webkit ? y : 0;
-
-    if (!hasScrollbarGutterStable) {
-      Object.assign(htmlStyle, {
-        position: 'fixed',
-        top: `${-scrollY + visualY}px`,
-        left: `${-scrollX + visualX}px`,
-        right: '0',
-      });
-    }
-
-    Object.assign(htmlStyle, {
+    Object.assign(html.style, {
       overflowY:
         !hasScrollbarGutterStable && (isScrollableY || hasConstantOverflowY) ? 'scroll' : 'hidden',
       overflowX:
         !hasScrollbarGutterStable && (isScrollableX || hasConstantOverflowX) ? 'scroll' : 'hidden',
     });
 
-    // Ensure two scrollbars can't appear since `<html>` now has a forced scrollbar, but the
-    // `<body>` may have one too.
-    if (isScrollableY || hasConstantOverflowY) {
-      bodyStyle.overflowY = 'visible';
-    }
-    if (isScrollableX || hasConstantOverflowX) {
-      bodyStyle.overflowX = 'visible';
-    }
-  }
+    // Avoid shift due to the default <body> margin. This does cause elements to be clipped
+    // with whitespace. Warn if <body> has margins?
+    const bodyVerticalMargins =
+      parseFloat(bodyStyles.marginTop) + parseFloat(bodyStyles.marginBottom);
 
-  function cleanup() {
-    if (isFirefox) {
-      Object.assign(bodyStyle, {
-        overflow: '',
-        [paddingProp]: '',
-      });
-      Object.assign(htmlStyle, {
-        overflow: '',
-      });
-      return;
-    }
+    Object.assign(body.style, {
+      position: 'relative',
+      height: `calc(100dvh - ${bodyVerticalMargins}px)`,
```

Comment:
A total turbo nit, but since in the vast majority of websites the margins will be reset to 0, can we check for that and if yes, set height to `100dvh`? Just so it's not `calc(100dvh - 0px)` when inspecting